### PR TITLE
[WIP][SPARK-55206][PYTHON][SQL] Minimal Python Transplantation

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

This adds basic experimental transplantation from Python UDFs into Catalyst. This replaces https://github.com/apache/spark/pull/53547

### Why are the changes needed?

Python UDF evaluation is slow and Pytthon UDF usage has increased with Pandas + Spark (prev Koalas prev Sparkling Pandas)

### Does this PR introduce _any_ user-facing change?

Yes new flag (defaults to falls)

### How was this patch tested?

Combination of regular unit tests and hypothesis testing tool to look for edge cases.

### Was this patch authored or co-authored using generative AI tooling?

Yes, Hypothesis testing was generated by gen AI, test refactor by claude,

### CODE REVIEW

File: python/pyspark/sql/udf.py [hunk 1]
```diff
@@ -206,6 +206,64 @@ def __init__(
...
+            )
+            self._transpile_errors.append("Transpilation only functions in ANSI mode.")
+            transpile_enabled = False
+        if transpile_enabled and session:
+            # Import only if needed, also avoid circular import loops.
+            from pyspark.sql.transpile import _transpile_func
+
+            try:
+                self.transpiled, errors, self._transpiled_param_names = _transpile_func(
+                    session, func, returnType
```
[the remark this answers is not in the document]
Comment (holdenk) [reply python/pyspark/sql/udf.py]: I mean yes it bypasess the Pandas/Arrow UDF path but is there a case where it would be semantically diffirent? I can't think of one off the top of my head.

File: python/pyspark/sql/udf.py [hunk 2]
```diff
@@ -421,7 +480,12 @@ def _create_judf(self, func: Callable[..., Any]) -> "JavaObject":
         jdt = spark._jsparkSession.parseDataType(self.returnType.json())
         assert sc._jvm is not None
         judf = getattr(sc._jvm, "org.apache.spark.sql.execution.python.UserDefinedPythonFunction")(
-            self._name, wrapped_func, jdt, self.evalType, self.deterministic
+            self._name,
+            wrapped_func,
+            jdt,
+            self.evalType,
+            self.deterministic,
+            map(_to_java_column_opt, self.transpiled),
```
[the remark this answers is not in the document]
Comment (holdenk) [reply python/pyspark/sql/udf.py]: Demonstratebly false based on the fact that the code path works.

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -676,6 +683,7 @@ jobs:
       METASPACE_SIZE: 1g
       BRANCH: ${{ inputs.branch }}
       PYSPARK_TEST_TIMEOUT: 450
+      RUN_HYPOTHESIS: ${{ fromJson(needs.precondition.outputs.required).transpile }}
```
[the remark this answers is not in the document]
Comment (holdenk) [reply .github/workflows/build_and_test.yml:688]: I don't think so we require _hypothesis_enabled to be set too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -989,6 +992,72 @@ object LimitPushDown extends Rule[LogicalPlan] {
...
+          s.pythonUDFExpr.mapChildren(applyExpr(_, parent_is_udf = true))
+        } else if (!conf.getConf(SQLConf.ATTEMPT_TRANSPILATION_OF_PYTHON_UDFS)) {
+          logWarning(
+            "Skipping Python UDF transpilation: " +
+              s"${SQLConf.ATTEMPT_TRANSPILATION_OF_PYTHON_UDFS.key} is disabled but " +
+              "we still got TranspiledPythonUDFs in our plan."
+          )
+          s.pythonUDFExpr.mapChildren(applyExpr(_, parent_is_udf = true))
+        } else if (!parent_is_udf ||
+          !s.children.forall { x => x.isInstanceOf[PythonUDF] }) {
```
[the remark this answers is not in the document]
Comment (holdenk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Hmm we actually probably need to rework this guard since the child could be a transpiled or non-transpiled one and then it depends on how that collapses and because we work top down in the tree we don't know that yet.
